sdk/go/common/resource/plugin/provider-implementers-guide.md

```diff
@@ -0,0 +1,609 @@
+successful execution of a Pulumi progam in the resource's stack, the engine deletes the
+resource by calling the resource's provider's [`Delete`](#delete) method with the resource's ID and
+last refereshed state.
+
+The diagram below summarizes the custom resource lifecycle. Detailed descriptions of each
+resource operation follow.
+
+![Custom Resource Lifeycle Diagram](./resource_lifecycle.svg)
+
+### Lifecycle Methods
+
+#### Check
+
+The `Check` method is responsible for validating the inputs to a resource. It may
+optionally apply default values for unspecified input properties that cannot reasonably
+be computed outside the provider (e.g. because they require access to the provider's
+internal data structures).
+
+##### Parameters
+
+- `urn`: the [URN](#urns) of the resource.
+- `olds`: the last recorded input `Object` for the resource, if any. If present, these
+          inputs must have been generated by a prior call to `Check` or [`Read`](#read).
+          These inputs will never contain [`Unknown`s](#unknowns).
+- `news`: the new input `Object` for the resource. These inputs may have been provided by
+          the user or generated by a call to [`Read`](#read), and may contain
+          [`Unknown`s](#unknowns).
+
+##### Results
+
+- `inputs`: the checked input `Object` for the resource with default values applied. The
+            types of the properties in `inputs` should agree with the types of the
+            resource's input properties as described in its (schema)[#schema]. If `news`
+            contains [`Unknown`s](#unknowns), `inputs` may contain [`Unknown`s](#unknowns).
+- `failures`: any validation failures present in the inputs. These failures should be
+              constrained to type and range mismatches. A failure is a tuple of a
+              [property path](#property-paths) and a failure reason.
+
+#### Diff
+
+The `Diff` method is responsible for calculating the differences between the actual and
+desired state of a resource as represented by its last recorded state and new input
+`Object` as returned from [`Check`](#check) or [`Read`](#read) and the logical
+operation necessary to reconcile the two (i.e. no operation, an `Update, or a `Replace`).
+
+##### Parameters
+
+- `urn`: the [URN](#urns) of the resource.
+- `id`: the [ID](#custom-resources) of the resource.
+- `olds`: the last recorded state `Object` for the resource. This `Object` must have been
+          generated by a call to `Create`, `Read`, or `Update`, and will never contain
+          [`Unknown`s](#unknowns).
+- `news`: the current input `Object` for the resource as returned by [`Check`](#check) or
+          [`Read`](#read). This value may contain [`Unknown`s](#unknowns).
+- `ignoreChanges`: the set of [property paths](#property-paths) to treat as unchanged.
+
+##### Results
+
+- `detailedDiff`: the [detailed diff](#detailed-diffs) between the resource's actual and
```

Comment:
This is still optional - right?

Worth noting what happens if you don't return it?  Or are you intentionally trying to lean toward insisting this is returned?